xrmocap/service/smpl_verts_service.py

```diff
@@ -168,122 +164,156 @@ def on_disconnect(self) -> None:
-        """Call body_model.forward() to get SMPL vertices.
+            Args:
+                data (dict): Request data, frame_idx is required.
 
-        Args:
-            data (dict): Request data, frame_idx is required.
+            Returns:
+                dict: Response data.
+                    If success, status is 'success' and
+                    vertices bytes for an ndarray.
+            """
+            resp_dict = dict()
+            req_dict = data
+            uuid_str = session['uuid']
+            frame_idx = req_dict['frame_idx']
+            smpl_data = session['smpl_data']
+            # check if data and args are valid
+            failed = False
+            if smpl_data is None:
+                error_msg = f'Client {uuid_str}\'s smpl data not uploaded.'
+                failed = True
+            elif frame_idx >= smpl_data.get_batch_size():
+                error_msg = f'Client {uuid_str}\'s smpl data only has ' +\
+                    f'{smpl_data.get_batch_size()} frames, ' +\
+                    f'but got frame_idx={frame_idx} in request.'
+                failed = True
+            if failed:
+                self.logger.error(error_msg)
+                resp_dict['msg'] = f'Error: {error_msg}'
+                resp_dict['status'] = 'fail'
+                return resp_dict
+            # no error, forward body model
+            else:
+                tensor_dict = smpl_data.to_tensor_dict(
+                    repeat_betas=True, device=self.device)
+                for k, v in tensor_dict.items():
+                    tensor_dict[k] = v[frame_idx:frame_idx + 1]
+                body_model = session['body_model']
+                with self.worker_lock:
+                    self.forward_timer.start()
+                    with torch.no_grad():
+                        body_model_output = body_model(**tensor_dict)
+                    self.forward_timer.stop()
+                    if self.forward_timer.count >= 50:
+                        self.logger.info(
+                            'Average forward time per-frame:' +
+                            f' {self.forward_timer.get_average(reset=True):.4f} s')
+                verts = body_model_output['vertices']  # n_batch=1, n_verts, 3
+                verts_np = verts.cpu().numpy().squeeze(0).astype(np.float16)
+                session['last_connect_time'] = time.time()
+                if self.enable_bytes:
+                    verts_bytes = verts_np.tobytes()
+                    if self.enable_gzip:
+                        verts_bytes = gzip.compress(verts_bytes)
+                    return verts_bytes
+                else:
+                    resp_dict['verts'] = verts_np.tolist()
+                    resp_dict['status'] = 'success'
+                    return resp_dict
+        
+        @self.socketio.on('get_faces')
+        def get_faces() -> dict:
+            resp_dict = dict()
+            faces = session['faces']
```

Comment:
faces is included in the body model, it is not necessary to save it in session again. 
```suggestion
            body_model = session['body_model']
            if body_model is None:
                # log and respond
            else:
                faces_np = body_model.faces.cpu().numpy()
```

Reply:
1. Removed session['faces'].
2. body_model.faces is already a ndarray, therefore, faces.cpu().numpy() is no longer needed.